posts/2018-08-07-refactoring-monads.md

```diff
@@ -0,0 +1,391 @@
+    case None => Failure(new ParseError(formatStr))
+    case Some(Print) =>
+      Success(PrintBook(id, title, author))
+    case Some(Digital) =>
+      extractEBook(id, title, author, downloadTypeStrOpt)
+  }
+
+def findBookById(id: Int): Try[Book] =
+  DB.unsafeQueryUnique(sql"""select * from catalog where id = $id""").flatMap { row =>
+    val id = row[Int]("id")
+    val title = row[String]("title")
+    val author = row[String]("author")
+    val formatStr = row[String]("format")
+    val downloadTypeStr = row[Option[String]]("download_type")
+    extractBook(id, title, author, formatStr, downloadTypeStr)
+  }
+```
+
+As you can see, this form has more manageably-sized functions,
+although they are still a little long. You can also see that the flow
+of control is distributed through all three functions, which means
+understanding the logic enough to modify or test it requires
+understanding all three functions both individually and as a whole. To
+follow the logic, we must trace the functions like a recursive descent
+parser.
+
+### Refactoring with Monads
+Without throwing exceptions and catching them at the top, it's going
+to be hard to do substantially better than the "tail-refactoring"
+approach, unless we start to make use of the fact that we're working
+with `Try`, a data type that supports `flatMap`. More precisely, `Try`
+has a monad - recall that monads let us model computational effects
+that take place in sequence.
+
+Let's try to factor out smaller functions, each returning `Try`, and
+then use a for-comprehension to specify the sequence of operations:
+
+```tut:silent
+import scala.util.{Failure, Success, Try}
+
+def tryParse[A](s: String, parse: String => Option[A]): Try[A] = {
+  parse(s)
+    .map(Success(_))
+    .getOrElse(Failure(new ParseError(s)))
+}
+
+def parseFormat(s: String): Try[Format] = tryParse(s, Format.fromString)
+
+def parseDownloadType(o: Option[String], id: Int): Try[DownloadType] = {
+  o.map(tryParse(_, DownloadType.fromString))
+    .getOrElse(Failure(new AssertionError(s"download type not provided for digital book $id")))
+}
+
+def findBookById(id: Int): Try[Book] =
+  for {
+    row <- DB.unsafeQueryUnique(sql"""select * from catalog where id = $id""")
+    format <- parseFormat(row[String]("format"))
+    (id, title, author) = (row[Int]("id"), row[String]("title"), row[String]("author"))
+    book <- format match {
+      case Print =>
+        Success(PrintBook(id, title, author))
+      case Digital =>
+        parseDownloadType(row[Option[String]]("download_type"), id)
+          .map(EBook(id, title, author, _))
+    }
+  } yield book
+```
+
+It's less code, the functions are smaller, and the top-level function
+dictates the entire flow of control. No function takes more than 2
+arguments. These are testable, understandable functions. This version
+really shows the power of using Monads to sequence computation.
+
+Now we are truly making use of the fact that `Try` has a monad and not
```

Comment:
again, "monad instance"

Reply:
Thanks. I'm never sure about how to write that, but "has a monad instance" seems right. I'm not a category theory expert, but it always feels wrong to say "X is a monad" but rather "there is a monad for X". By the definition, it seems like a monad is really something that exists outside of, but in conjunction with whatever X is.